tests/utils/test_safe_eval.py

```
@@ -0,0 +1,55 @@
+    ("1+2+x", None, {"x": 4}, 7),
+    ("1*2+x", {"x": 4}, None, 6),
+    ("(1+2)*3", None, None, 9),
+    ("foo+bar", {"foo": 1030}, {"bar": 204}, 1234),
+]
+
+BAD_EXPRS = [("foo()",), ("foo.bar",), ("foo[123]",), ("(1,2)",), ("[3,4]",), ("int.__class__.__init__.__globals__",)]
+
+
+class TestSafeEval(unittest.TestCase):
+    @parameterized.expand(GOOD_EXPRS)
+    def test_good_exprs(self, expr, globals, locals, expected):
```

Ruff (0.15.17)

[error] Function argument `globals` is shadowing a Python builtin

(A002)